src/dd/dd_enc.c

```diff
@@ -0,0 +1,288 @@
+/**
+ * @file dd_enc.c Dependency Descriptor (DD) -- encoder
+ *
+ * Copyright (C) 2023 Alfred E. Heggestad
+ */
+
+#include <string.h>
+#include <re.h>
+#include <re_dd.h>
+
+
+#define DEBUG_MODULE "dd"
+#define DEBUG_LEVEL 5
+#include <re_dbg.h>
+
+
+static int mandatory_descriptor_fields(struct putbit *pb, const struct dd *dd)
+{
+	int err = 0;
+
+	err |= putbit_one(pb, dd->start_of_frame);
+	err |= putbit_one(pb, dd->end_of_frame);
+	err |= putbit_write(pb, 6, dd->frame_dependency_template_id);
+	err |= putbit_write(pb, 16, dd->frame_number);
+
+	return err;
+}
+
+
+static uint8_t next_layer(const struct dd *dd, unsigned prev, unsigned next)
+{
+	if (dd->template_spatial_id[next] == dd->template_spatial_id[prev] &&
+	    dd->template_temporal_id[next] == dd->template_temporal_id[prev]) {
+
+		return DD_SAME_LAYER;
+	}
+	else if (dd->template_spatial_id[next] ==
+		 dd->template_spatial_id[prev] &&
+		 dd->template_temporal_id[next] ==
+		 dd->template_temporal_id[prev] + 1) {
+
+		return DD_NEXT_TEMPORAL_LAYER;
+	}
+	else if (dd->template_spatial_id[next] ==
+		 dd->template_spatial_id[prev] + 1 &&
+		 dd->template_temporal_id[next] == 0) {
+
+		return DD_NEXT_SPATIAL_LAYER;
+	}
+
+	return DD_NO_MORE_TEMPLATES;
+}
+
+
+static int template_layers(struct putbit *pb, const struct dd *dd)
+{
+	int err = 0;
+
+	for (unsigned i = 1; i < dd->template_cnt; ++i) {
+
+		uint8_t next_layer_idc = next_layer(dd, i - 1, i);
+		if (next_layer_idc == DD_NO_MORE_TEMPLATES)
+			return EBADMSG;
+
+		err |= putbit_write(pb, 2, next_layer_idc);
+	}
+
+	/* end of layers */
+	err |= putbit_write(pb, 2, DD_NO_MORE_TEMPLATES);
+
+	return err;
+}
+
+
+static int template_dtis(struct putbit *pb, const struct dd *dd)
+{
+	re_printf("template_dtis\n");
+
+	for (uint8_t templateIndex = 0;
+	     templateIndex < dd->template_cnt;
+	     templateIndex++) {
+
+		for (uint8_t dtIndex = 0; dtIndex < dd->dt_cnt; dtIndex++) {
+
+			/* See table A.1 below for meaning of DTI values. */
+
+			uint8_t v = dd->template_dti[templateIndex][dtIndex];
+
+			int err = putbit_write(pb, 2, v);
+			if (err)
+				return err;
+		}
+	}
+
+	return 0;
+}
+
+
+static int template_fdiffs(struct putbit *pb, const struct dd *dd)
+{
+	int err;
+
+	re_printf("template_fdiffs (encode) bit_pos=%zu\n", pb->bit_pos);
```

Comment:
should this `DEBUG_INFO` instead of `re_printf`?

Reply:
I removed the logging instead ...